common/archiver/gcloud/historyArchiver.go

```diff
@@ -0,0 +1,408 @@
+	}
+	return nil, err
+}
+
+func newHistoryArchiver(container *archiver.HistoryBootstrapContainer, historyIterator archiver.HistoryIterator, storage connector.Client) archiver.HistoryArchiver {
+	return &historyArchiver{
+		container:       container,
+		gcloudStorage:   storage,
+		historyIterator: historyIterator,
+	}
+}
+
+// Archive is used to archive a workflow history. When the context expires the method should stop trying to archive.
+// Implementors are free to archive however they want, including implementing retries of sub-operations. The URI defines
+// the resource that histories should be archived into. The implementor gets to determine how to interpret the URI.
+// The Archive method may or may not be automatically retried by the caller. The ArchiveOptions are used
+// to interact with these retries including giving the implementor the ability to cancel retries and record progress
+// between retry attempts.
+// This method will be invoked after a workflow passes its retention period.
+func (h *historyArchiver) Archive(ctx context.Context, URI archiver.URI, request *archiver.ArchiveHistoryRequest, opts ...archiver.ArchiveOption) (err error) {
+	scope := h.container.MetricsClient.Scope(metrics.HistoryArchiverScope, metrics.DomainTag(request.DomainName))
+	featureCatalog := archiver.GetFeatureCatalog(opts...)
+	sw := scope.StartTimer(metrics.CadenceLatency)
+	defer func() {
+		sw.Stop()
+		if err != nil {
+
+			if err.Error() == errUploadNonRetriable.Error() {
+				scope.IncCounter(metrics.HistoryArchiverArchiveNonRetryableErrorCount)
+			}
+
+			if featureCatalog.NonRetriableError != nil {
+				err = featureCatalog.NonRetriableError()
+			}
+
+		}
+	}()
+
+	logger := archiver.TagLoggerWithArchiveHistoryRequestAndURI(h.container.Logger, request, URI.String())
+
+	if err := h.validateURI(URI); err != nil {
+		logger.Error(archiver.ArchiveNonRetriableErrorMsg, tag.ArchivalArchiveFailReason(archiver.ErrReasonInvalidURI), tag.Error(err))
+		return errUploadNonRetriable
+	}
+
+	if err := archiver.ValidateHistoryArchiveRequest(request); err != nil {
+		logger.Error(archiver.ArchiveNonRetriableErrorMsg, tag.ArchivalArchiveFailReason(archiver.ErrReasonInvalidArchiveRequest), tag.Error(err))
+		return errUploadNonRetriable
+	}
+
+	var totalUploadSize int64
+	historyIterator := h.historyIterator
+	var progress progress
+	if historyIterator == nil { // will only be set by testing code
+		historyIterator, _ = loadHistoryIterator(ctx, request, h.container.HistoryV2Manager, featureCatalog, &progress)
+	}
+
+	for historyIterator.HasNext() {
+		part := progress.CurrentPageNumber
+		historyBlob, err := getNextHistoryBlob(ctx, historyIterator)
+
+		if err != nil {
+			logger := logger.WithTags(tag.ArchivalArchiveFailReason(archiver.ErrReasonReadHistory), tag.Error(err))
+			if !common.IsPersistenceTransientError(err) {
+				logger.Error(archiver.ArchiveNonRetriableErrorMsg)
+			} else {
+				logger.Error(archiver.ArchiveTransientErrorMsg)
```

Comment:
If it's a transient error, don't return `errUploadNonRetriable`

Reply:
Done